tests/btb/recommendation/test_recommender.py

```diff
@@ -49,19 +51,34 @@ def test__get_candidates_all(self):
 
     def test__get_candidates_some(self):
         recommender = BaseRecommender(self.dpp_matrix)
-        X = np.array([0, 0, 0, 1, 0, 1, 0, 1, 0, 1, 0, 0, 0, 1, 0, 0])
-        recommender.dpp_vector = X
+        #X = np.array([0, 0, 0, 1, 0, 1, 0, 1, 0, 1, 0, 0, 0, 1, 0, 0])
+        #recommender.dpp_vector = X
+        recommender.tried_pipelines = set([3, 5, 7, 9, 13])
         expected = np.array([0, 1, 2, 4, 6, 8, 10, 11, 12, 14, 15])
         candidates = recommender._get_candidates()
         np.testing.assert_array_equal(
             candidates,
             expected,
         )
 
+    def test__get_candidates_with_add_zeros(self):
+        recommender = BaseRecommender(self.dpp_matrix)
+        recommender.add({0: 0, 2: 0})
+        expected = np.array(
+            [1, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15]
+        )
+        candidates = recommender._get_candidates()
+        np.testing.assert_array_equal(
+            candidates,
+            expected,
+        )
+        assert set([0, 2]) == recommender.tried_pipelines
```

Comment:
always prefer the unittest.TestCase assertion methods because they provide more debugging information and flexibility
```python
expected_tried_pipelines = set([0, 2])
self.assertEqual(expected_tried_pipelines, recommender.tried_pipelines)
```

See https://docs.python.org/3/library/unittest.html#assert-methods for more details